File: Doubao_MarsCode/go/MC0035/main.go
package main

import (
	"fmt"
)

func solution(a []int, b []int) []int {
	if len(a) == 0 || len(b) == 0 {
		return []int{}
	}

	m := make(map[int]int)
	for _, v := range a {
		m[v]++
	}

	res := make([]int, 0)
	for i := len(b) - 1; i >= 0; i-- {
		if m[b[i]] > 0 {
			res = append(res, b[i])
		}
	}
	return res
}

func main() {
	fmt.Println(equal(solution([]int{1, 2, 3, 7}, []int{2, 5, 7}), []int{7, 2}))
	fmt.Println(equal(solution([]int{1, 4, 8, 10}, []int{2, 4, 8, 10}), []int{10, 8, 4}))
	fmt.Println(equal(solution([]int{3, 5, 9}, []int{1, 4, 6}), []int{}))
	fmt.Println(equal(solution([]int{1, 2, 3}, []int{1, 2, 3}), []int{3, 2, 1}))
}

func equal(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: Doubao_MarsCode/java/MC0035/Main.java
package MC0035;

import java.util.*;

/**
 * 35. 小S的倒排索引
 */
public class Main {
    public static List<Integer> solution(List<Integer> a, List<Integer> b) {
        if (a.isEmpty() || b.isEmpty()) {
            return Collections.emptyList();
        }

        Set<Integer> set = new HashSet<>(a);
        List<Integer> res = new ArrayList<>();
        for (int i = b.size() - 1; i >= 0; i--) {
            if (set.contains(b.get(i))) {
                res.add(b.get(i));
            }
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(solution(Arrays.asList(1, 2, 3, 7), Arrays.asList(2, 5, 7)).equals(Arrays.asList(7, 2)));
        System.out.println(solution(Arrays.asList(1, 4, 8, 10), Arrays.asList(2, 4, 8, 10)).equals(Arrays.asList(10, 8, 4)));
        System.out.println(solution(Arrays.asList(3, 5, 9), Arrays.asList(1, 4, 6)).equals(Collections.emptyList()));
        System.out.println(solution(Arrays.asList(1, 2, 3), Arrays.asList(1, 2, 3)).equals(Arrays.asList(3, 2, 1)));
    }
}